File: solutions/elixir/strain/1/lib/strain.ex
defmodule Strain do
  @doc """
  Given a `list` of items and a function `fun`, return the list of items where
  `fun` returns true.

  Do not use `Enum.filter`.
  """
  @spec keep(list :: list(any), fun :: (any -> boolean)) :: list(any)
  def keep(list, fun) do
    keep_it([], list, fun)
  end

  defp keep_it(result, [], _), do: result
  defp keep_it(result, [fst | rst], f) do
    new_result = if f.(fst) do 
      result ++ [fst]
    else
      result
    end
    keep_it(new_result, rst, f)
  end

  @doc """
  Given a `list` of items and a function `fun`, return the list of items where
  `fun` returns false.

  Do not use `Enum.reject`.
  """
  @spec discard(list :: list(any), fun :: (any -> boolean)) :: list(any)
  def discard(list, fun) do
    keep(list, fn x -> !fun.(x) end)
  end
end

File: solutions/scala/strain/1/src/main/scala/Strain.scala
object Strain {
  def keep[A](input: Seq[A], predicate: A => Boolean): Seq[A] =
    input.foldRight(Seq.empty[A]) { (item, kept) =>
      if (predicate(item)) item +: kept else kept
    }

  def discard[A](input: Seq[A], predicate: A => Boolean): Seq[A] =
    keep(input, item => !predicate(item))
}
